File: app/src/main/java/dev/yiray/qickmvivm/base/BaseState.java
package dev.yiray.qickmvivm.base;

public interface BaseState<VS extends BaseState.BaseViewState<VS>> {
    VS toViewState();

    abstract class BaseViewState<VS extends BaseViewState<VS>> implements BaseState<VS> {

        abstract class Builder {

            Builder () {

            }

            public abstract VS build();
        }

        public abstract VS copy(VS nextState);
    }

    interface SideEffect {

    }

}

File: app/src/main/java/dev/yiray/livedatamvi/base/BaseState.kt
package dev.yiray.livedatamvi.base

import dev.yiray.livedatamvi.base.BaseState.BaseViewState

interface BaseState<VS : BaseViewState<VS>?> {
    fun toViewState(): VS

    abstract class BaseViewState<VS : BaseViewState<VS>?> : BaseState<VS> {
        abstract fun copy(nextState: VS): VS
    }

    interface SideEffect
}